Initial commit of PGVector documentation (#5380)
voxel51/fiftyone: docs/source/integrations/pgvector.rst

.. pgvector-integration:

PostgreSQL Vector Search Integration

> =================================

Reviewer: nit: Sphinx emits errors if the `===` don't exactly match the length of the text above them

.. default-role:: code

> `PostgreSQL <https://www.postgresql.org>`_ is the world's most advanced open source database, and we've made it easy to use PostgreSQL's `vector search capabilities using PGVector <https://github.com/pgvector/pgvector>`_ on your computer vision data directly from FiftyOne!

Reviewer: nit: our [docs style guide](https://github.com/voxel51/fiftyone/blob/develop/STYLE_GUIDE.md#documentation-style-guide) requests that maximum line length is 79 characters.

Unfortunately, enforcing this via [pre-commit hooks](https://github.com/voxel51/fiftyone/blob/09e7193a3953b03110a45fae42ebf154a0e1471f/.pre-commit-config.yaml#L8) causes bad things to happen (eg tables are broken, I believe), so the line length must be manually enforced via your `ENTER` key...

    model = foz.load_zoo_model(model_name)
    brain_key = "pgvector_index"

    # Option 1: Compute embeddings on the fly from model name
    fob.compute_similarity(
        dataset,
        model=model_name,
        embedding_column="embedding_column",  # the column in which to store the embeddings
        backend="pgvector",
        brain_key=brain_key,
    )

    # Option 2: Compute embeddings on the fly from model instance
    fob.compute_similarity(
        dataset,
        model=model,
        embedding_column="embedding_column",  # the column in which to store the embeddings
        backend="pgvector",
        brain_key=brain_key,
    )

    # Option 3: Pass precomputed embeddings as a numpy array
    embeddings = dataset.compute_embeddings(model)
    fob.compute_similarity(
        dataset,
        embedding_column="embeddings",  # the column in which to store the embeddings
        backend="pgvector",
        brain_key=brain_key,
    )

    # Option 4: Pass precomputed embeddings by field name
    embeddings = dataset.compute_embeddings(model)
    dataset.set_values("embeddings", embeddings.tolist())
    fob.compute_similarity(
        dataset,

>         embedding_column="embeddings",  # the field that contains the embeddings

Reviewer: Does `embedding_column` really specify both the PGvector column to use *and* a pre-existing FiftyOne field from which to load pre-computed embeddings?

The [`fob.compute_similarity()` interface](https://github.com/voxel51/fiftyone-brain/blob/a2922cb3510d3087e64c8b36fc826bb32ab76c42/fiftyone/brain/__init__.py#L543) already mandates that the `embeddings` parameter should be usable for all backends to load pre-existing embeddings and/or persist embeddings that are computed by the backend.

So the syntax I would expect, albeit a bit redundant, would be:

```py
fob.compute_similarity(
    dataset,
    embeddings="embeddings",  # the FO field that contains the embeddings
    embedding_column="embeddings",  # the PGvector column in which to store the embeddings
    backend="pgvector",
    brain_key=brain_key,
)
```

Reviewer 2: @brimoor Yes - will do after the engineering code review and test suite PR

    :linenos:

    import fiftyone as fo
    import fiftyone.brain as fob
    import fiftyone.zoo as foz

    # Step 1: Load your data into FiftyOne
    dataset = foz.load_zoo_dataset("quickstart")

    # Steps 2 and 3: Compute embeddings and create a similarity index
    pgvector_index = fob.compute_similarity(
        dataset, 
        embeddings="embeddings",  # the field in which to store the embeddings
        brain_key="pgvector_index",
        backend="pgvector",
    )

Once the similarity index has been generated, we can query our data in FiftyOne by specifying the `brain_key`:

.. code-block:: python
    :linenos:

    # Step 4: Query your data
    query = dataset.first().id  # query by sample ID
    view = dataset.sort_by_similarity(
        query, 
        brain_key="pgvector_index",
        k=10,  # limit to 10 most similar samples
    )

    # Step 5 (optional): Cleanup

    # Delete the PGVector index
    pgvector_index.cleanup()

    # Delete run record from FiftyOne
    dataset.delete_brain_run("pgvector_index")

.. note::

    Skip to :ref:`this section <pgvector-examples>` for a variety of common PGVector query patterns.

.. _postgresql-setup:

Setup
_____

> In order to get started, you must set up both a PostgreSQL instance and PGVector alongside it.

Reviewer 2: ```suggestion
In order to get started, you must set up both a PostgreSQL instance with the PGVector extension installed.
```
